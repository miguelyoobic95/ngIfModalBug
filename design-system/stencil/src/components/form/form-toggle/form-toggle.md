---
name: Toggle
category: Form Components
---

## CSS

|Type|Name|Description|
|---|---|---|
|Background Fill|`background-fill`|fills the background of the toggle|
|Accent|`accent`|yoo-form-toggle accent|
|Dark|`dark`|yoo-form-toggle dark|
|Success|`success`|yoo-form-toggle success|
|Danger|`danger`|yoo-form-toggle danger|
|Info|`info`|yoo-form-toggle info|
|Warning|`warning`|yoo-form-toggle warning|


```yoo-form-toggle.html
<div class="container">
    <div class="column-container">
        <yoo-form-toggle class="accent"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="dark"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="success"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="danger"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="info"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="warning"></yoo-form-toggle>
    </div>
    <div class="column-container">
        <yoo-form-toggle class="accent background-fill"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="dark background-fill"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="success background-fill"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="danger background-fill"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="info background-fill"></yoo-form-toggle>
        <br>
        <yoo-form-toggle class="warning background-fill"></yoo-form-toggle>
    </div>
</div>

```

```yoo-form-toggle.css hidden
    .container {
        display: flex;
    }

    .column-container {
        padding-right: 10px;
        display: flex;
        flex-direction: column;
    }
    
```

## Events

|Attr|Description|
|---|---|---|
|`toggled`|boolean event emitted on toggle|