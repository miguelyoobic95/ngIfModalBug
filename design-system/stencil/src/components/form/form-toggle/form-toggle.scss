:host {
    .outer-container {
        width: 100%;
        .text-container {
            flex: 1;            
        }
        .toggle-container {
            height: 1.3125rem;
            width: 2.1875rem;
            border-radius: 0.96875rem;
            border: solid 1px rgba(0, 0, 0, 0.1);
            display: flex;
            align-items: center;
            &.active {
                border-color: cssvar(success);
                background-color: cssvar(success);
            }
            &:hover {
                cursor: pointer;
            }
            .inner-container {
                height: 1.25rem;
                width: 1.25rem;
                border-radius: 50%;
                background-color: cssvar(light);
                box-sizing: content-box;
                border: solid 0.5px rgba(0, 0, 0, 0.1);   
                box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.05), 0 2px 2px 0 rgba(0, 0, 0, 0.1), 0 1px 1px 0 rgba(0, 0, 0, 0.05);
                transform: translateX(-1px);                   
                &:hover {
                    cursor: pointer;
                }
                &.active {
                    border: solid 1px cssvar(success);
                    height: 1.1875rem;
                    width: 1.1875rem;
                    transform: translateX(14px);                   
                }
            }
        }
    }
}

:host(.line) {
    .outer-container {
        .text-container {
            order: 1;
            margin-left: 1rem;
        }
    }
}

@mixin class-color-applicator($class, $color) {
    :host(.#{$class}) {
        .outer-container {
            .toggle-container {
                &.active {
                    border-color: cssvar($color);
                    background-color: cssvar($color);
                }
                .inner-container {
                    &.active {
                        border-color: cssvar($color);
                    }
                }
            }
        }
    }
}

@include class-color-applicator(accent, accent);
@include class-color-applicator(success, success);
@include class-color-applicator(danger, danger);
@include class-color-applicator(info, info);
@include class-color-applicator(warning, warning);
@include class-color-applicator(dark, dark);