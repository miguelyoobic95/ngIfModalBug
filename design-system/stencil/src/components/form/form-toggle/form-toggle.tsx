import { Component, Element, Event, EventEmitter, Prop } from '@stencil/core';
import { setAnimation, animations } from '../../../utils/anim';
import { setValidator, setAsyncValidator, setValueAndValidateInput } from '../../../utils/helpers/form-input-helpers';
import { IFormToggle, FormToogleType, Validator, AsyncValidator, ValidatorEntry } from '@shared/interfaces';

@Component({
    tag: 'yoo-form-toggle',
    styleUrl: 'form-toggle.scss',
    scoped: true
})
export class YooFormToggleComponent implements IFormToggle {

    @Prop({ mutable: true }) value: boolean;
    @Prop() validators: Array<Validator<boolean> | ValidatorEntry> = [];
    @Prop() asyncValidators: Array<AsyncValidator<boolean>>;
    @Prop() readonly: boolean;
    @Prop() type: FormToogleType = 'normal';
    @Prop() text: string;

    @Event() validityChanged: EventEmitter<boolean>;
    @Event() inputBlurred: EventEmitter<any>;
    @Event() inputFocused: EventEmitter<boolean>;
    @Event() inputChanged: EventEmitter<any>;

    @Element() host: HTMLStencilElement;

    // Reduced Validators
    _validator: Validator<string> = (x: string) => true;
    _asyncValidator: AsyncValidator<string> = async (x: string) => true;

    componentWillLoad() {
        setValidator(this.validators);
        setAsyncValidator(this.asyncValidators);
        if (this.type === 'line') {
            this.host.classList.add('line');
        }
    }

    onToggle(): void {
        if (!this.readonly) {
            setAnimation(animations.slideHorizontal, this.host.querySelector('.inner-container'), { open: (this.value ? false : true), distance: 14, start: -1 });
            this.value = !this.value;
            setValueAndValidateInput(this.value, this);
        }
    }

    renderReadonly(): JSX.Element {
        return (
            <div class="readonly" attr-layout="row">
                {this.value === true ? <i class="yo-check-solid success"></i> : (this.value === false ? <i class="yo-cross-danger danger"></i> : <i class="yo-circle"></i>)}
            </div>
        );
    }

    renderEditable(): JSX.Element {
        return (
            <div class="outer-container" attr-layout="row">
                <div class="text-container">
                    <span>{this.text}</span>
                </div>
                <div class={'toggle-container' + (this.value ? ' active' : '')} attr-layout="row" onClick={() => this.onToggle()}>
                    <div class={'inner-container' + (this.value ? ' active' : '')}></div>
                </div>
            </div>
        );
    }

    render(): JSX.Element {
        return this.readonly ? this.renderReadonly() : this.renderEditable();
    }

}