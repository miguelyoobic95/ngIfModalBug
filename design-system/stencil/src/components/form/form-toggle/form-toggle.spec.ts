import { TestWindow } from '@stencil/core/testing';
import { YooFormToggleComponent } from './form-toggle';

jest.mock('../../../utils/anim');

fdescribe('YooFormToggleComponent', () => {
    let window;
    beforeEach(() => {
        window = new TestWindow();
    });
    it('should build', () => {
        expect(new YooFormToggleComponent()).toBeTruthy();
    });

    describe('Snapshots', () => {
        it('should render', async () => {
            let element;
            element = await window.load({
                components: [YooFormToggleComponent],
                html: '<yoo-form-toggle class="accent"></yoo-form-toggle>'
            });
            expect(element).toMatchSnapshot();
        });
    });

    describe('Events', async () => {
        it('should fire an event on click', async () => {
            let element = await window.load({
                components: [YooFormToggleComponent],
                html: '<yoo-form-toggle></yoo-form-toggle>'
            });
            let toggle: HTMLElement = element.querySelector('.outer-container');
            let res;
            let emitItem = jest.fn(ev => {res = ev.detail; });
            window.document.addEventListener('inputChanged', emitItem);
            await toggle.click();
            expect(emitItem).toHaveBeenCalled();
        });
    });
});
