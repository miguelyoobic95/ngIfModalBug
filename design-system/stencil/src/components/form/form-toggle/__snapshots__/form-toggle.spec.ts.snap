// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooFormToggleComponent Snapshots should render 1`] = `
<yoo-form-toggle
  class="accent hydrated"
  data-yoo-form-toggle-host=""
>
  
  <div
    attr-layout="row"
    class="outer-container"
    data-yoo-form-toggle=""
  >
    <div
      class="text-container"
      data-yoo-form-toggle=""
    >
      <span
        data-yoo-form-toggle=""
      >
        
      </span>
    </div>
    <div
      attr-layout="row"
      class="toggle-container"
      data-yoo-form-toggle=""
    >
      <div
        class="inner-container"
        data-yoo-form-toggle=""
      />
    </div>
  </div>
</yoo-form-toggle>
`;
